File: skills/steward-apple/scripts/apple.swift
import Foundation
import EventKit

struct Failure: Error, CustomStringConvertible { let description: String; init(_ s: String) { description = s } }
let store = EKEventStore()
let iso = ISO8601DateFormatter()
func emit(_ value: Any) { let data = try! JSONSerialization.data(withJSONObject: value, options: [.sortedKeys]); print(String(data: data, encoding: .utf8)!) }
func str(_ r: [String: Any], _ key: String) throws -> String {
    guard let s = r[key] as? String, !s.isEmpty else { throw Failure("Missing string: \(key)") }; return s
}
func date(_ s: String) throws -> Date {
    guard let d = iso.date(from: s) else { throw Failure("Use ISO8601 timestamp with explicit UTC offset: \(s)") }; return d
}
func waitFor(_ done: () -> Bool) throws {
    let deadline = Date().addingTimeInterval(90)
    while !done() && Date() < deadline { RunLoop.current.run(until: Date().addingTimeInterval(0.05)) }
    if !done() { throw Failure("Timed out; inspect permission or operation state before retrying") }
}
func access(_ type: EKEntityType, request: Bool) throws {
    if EKEventStore.authorizationStatus(for: type) == .fullAccess { return }
    guard request else { throw Failure("Access missing. Run connect for this app and approve the macOS prompt.") }
    var done = false; var granted = false; var problem: Error?
    let cb: (Bool, Error?) -> Void = { ok, e in granted = ok; problem = e; done = true }
    if type == .event { store.requestFullAccessToEvents(completion: cb) }
    else { store.requestFullAccessToReminders(completion: cb) }
    try waitFor { done }
    guard granted else { throw Failure("Access not granted: \(problem?.localizedDescription ?? "Check System Settings > Privacy & Security")") }
}
// Fail closed: macOS full access is broader than the user's write policy.
func checkWritePolicy(_ app: String, _ id: String) throws {
    let url = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support/Steward/apple-policy.json")
    guard let data = try? Data(contentsOf: url),
          let policy = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let apps = policy[app] as? [String: Any],
          let entry = apps[id] as? [String: Any], entry["write"] as? Bool == true else {
        throw Failure("Write blocked by Steward's private-container policy. Shared and unknown containers are read-only.")
    }
    guard let stamp = entry["privateVerifiedAt"] as? String, let verified = iso.date(from: stamp),
          Date().timeIntervalSince(verified) >= -60, Date().timeIntervalSince(verified) <= 900 else {
        throw Failure("Private status needs a fresh Calendar/Reminders UI check; refresh privateVerifiedAt only after verifying this container is not shared.")
    }
}
func calendar(_ id: String, _ type: EKEntityType, write: Bool = false) throws -> EKCalendar {
    guard let c = store.calendars(for: type).first(where: { $0.calendarIdentifier == id }) else { throw Failure("Calendar/list ID not found; rediscover containers") }
    if write && !c.allowsContentModifications { throw Failure("Container is read-only") }; return c
}
func eventJSON(_ e: EKEvent) -> [String: Any] {
    return ["id": e.eventIdentifier ?? "", "calendar": e.calendar.calendarIdentifier,
            "title": e.title ?? "", "start": iso.string(from: e.startDate), "end": iso.string(from: e.endDate),
            "allDay": e.isAllDay, "notes": e.notes ?? "", "location": e.location ?? "",
            "recurring": e.hasRecurrenceRules || e.isDetached, "attendees": e.attendees?.count ?? 0,
            "alarms": (e.alarms ?? []).map { ["relativeSeconds": $0.relativeOffset, "absolute": $0.absoluteDate.map { iso.string(from: $0) } ?? ""] as [String: Any] },
            "availability": e.availability.rawValue, "status": e.status.rawValue]
}
func reminderJSON(_ r: EKReminder) -> [String: Any] {
    var result: [String: Any] = ["id": r.calendarItemIdentifier, "calendar": r.calendar.calendarIdentifier,
      "title": r.title ?? "", "notes": r.notes ?? "", "completed": r.isCompleted, "priority": r.priority, "recurring": r.hasRecurrenceRules, "alarms": (r.alarms ?? []).map { ["relativeSeconds": $0.relativeOffset, "absolute": $0.absoluteDate.map { iso.string(from: $0) } ?? ""] as [String: Any] }]
    if let dc = r.dueDateComponents {
        result["dueComponents"] = ["year": dc.year ?? 0, "month": dc.month ?? 0, "day": dc.day ?? 0,
                                    "hour": dc.hour ?? -1, "minute": dc.minute ?? -1]
        if let d = Calendar.current.date(from: dc) { result["due"] = iso.string(from: d) }
    }
    return result
}
func reminders(_ c: EKCalendar) throws -> [EKReminder] {
    var result: [EKReminder]?; var done = false
    store.fetchReminders(matching: store.predicateForReminders(in: [c])) { result = $0; done = true }
    try waitFor { done }; guard let result else { throw Failure("Reminders fetch failed") }; return result
}
func event(_ r: [String: Any], _ c: EKCalendar) throws -> EKEvent {
    guard let e = store.event(withIdentifier: try str(r, "id")), e.calendar.calendarIdentifier == c.calendarIdentifier else { throw Failure("Event not found in selected calendar") }
    return e
}
func reminder(_ r: [String: Any], _ c: EKCalendar) throws -> EKReminder {
    guard let item = store.calendarItem(withIdentifier: try str(r, "id")) as? EKReminder,
        item.calendar.calendarIdentifier == c.calendarIdentifier else { throw Failure("Reminder not found in selected list") }; return item
}
func run(_ r: [String: Any]) throws -> [String: Any] {
    let op = try str(r, "op")
    if op == "status" {
        return ["calendarAuthorization": EKEventStore.authorizationStatus(for: .event).rawValue,
                "remindersAuthorization": EKEventStore.authorizationStatus(for: .reminder).rawValue,
                "timezone": TimeZone.current.identifier]
    }
    let app = try str(r, "app")
    guard ["calendar", "reminders"].contains(app) else { throw Failure("app must be calendar or reminders") }
    let type: EKEntityType = app == "calendar" ? .event : .reminder
    try access(type, request: op == "connect")
    if ["connect", "containers"].contains(op) {
        return ["containers": store.calendars(for: type).map { ["id": $0.calendarIdentifier, "name": $0.title,
            "source": $0.source.title, "sourceID": $0.source.sourceIdentifier, "writable": $0.allowsContentModifications] as [String: Any] }]
    }
    if op == "create-container" {
        guard r["apply"] as? Bool == true else { return ["state": "preview", "request": r] }
        let policyURL = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support/Steward/apple-policy.json")
        let data = try Data(contentsOf: policyURL)
        var policy = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        guard policy["allowContainerCreation"] as? Bool == true else { throw Failure("Container creation is disabled by local policy") }
        guard let source = store.sources.first(where: { $0.sourceIdentifier == r["source"] as? String }) else { throw Failure("Choose a discovered source ID") }
        let title = try str(r, "name")
        guard !store.calendars(for: type).contains(where: { $0.title == title && $0.source.sourceIdentifier == source.sourceIdentifier }) else { throw Failure("Container name already exists; inspect it instead of duplicating") }
        let c = EKCalendar(for: type, eventStore: store); c.title = title; c.source = source
        try store.saveCalendar(c, commit: true)
        var appPolicy = policy[app] as? [String: Any] ?? [:]
        appPolicy[c.calendarIdentifier] = ["name": title, "write": true, "privateVerifiedAt": iso.string(from: Date()), "provenance": "created-private-by-Steward"]
        policy[app] = appPolicy
        try JSONSerialization.data(withJSONObject: policy, options: [.prettyPrinted, .sortedKeys]).write(to: policyURL, options: .atomic)
        return ["state": "created", "container": ["id": c.calendarIdentifier, "name": c.title, "source": c.source.title]]
    }
    let mutating = ["create", "update", "delete", "rename-container", "delete-container"].contains(op)
    if mutating && r["apply"] as? Bool == true { try checkWritePolicy(app, str(r, "calendar")) }
    let c = try calendar(str(r, "calendar"), type, write: mutating)
    if op == "delete-container" {
        guard r["apply"] as? Bool == true else { return ["state": "preview", "request": r] }
        if type == .event {
            guard store.calendars(for: .event).filter({ $0.title == c.title }).count == 1 else { throw Failure("Ambiguous calendar name; exact empty check unavailable") }
            let task = Process(); let output = Pipe(); let errors = Pipe()
            task.executableURL = URL(fileURLWithPath: "/usr/bin/osascript")
            task.arguments = ["-l", "JavaScript", "-e", "function run(argv) { var a=Application('Calendar'); var cs=a.calendars.whose({name:argv[0]})(); if(cs.length!==1) throw new Error('Ambiguous calendar'); return cs[0].events().length; }", c.title]
            task.standardOutput = output; task.standardError = errors
            try task.run(); task.waitUntilExit()
            let count = String(data: output.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
            guard task.terminationStatus == 0, count == "0" else { throw Failure("Calendar nonempty or empty-check unavailable; deletion blocked") }
        } else {
            guard try reminders(c).isEmpty else { throw Failure("List is not empty; bulk deletion requires a separate reviewed workflow") }
        }
        try store.removeCalendar(c, commit: true)
        guard store.calendar(withIdentifier: try str(r, "calendar")) == nil else { throw Failure("Container deletion read-back unverified") }
        return ["state": "deleted", "id": try str(r, "calendar")]
    }
    if op == "rename-container" {
        guard r["apply"] as? Bool == true else { return ["state": "preview", "request": r] }
        c.title = try str(r, "name"); try store.saveCalendar(c, commit: true)
        return ["state": "renamed", "container": ["id": c.calendarIdentifier, "name": c.title]]
    }
    if op == "list" {
        if type == .event {
            let start = try date(str(r, "start")), end = try date(str(r, "end"))
            guard end > start, end.timeIntervalSince(start) <= 93 * 86400 else { throw Failure("Query window must be positive and at most 93 days") }
            let events = store.events(matching: store.predicateForEvents(withStart: start, end: end, calendars: [c]))
            return ["items": events.sorted { $0.startDate < $1.startDate }.map(eventJSON)]
        }
        let include = r["includeCompleted"] as? Bool ?? false
        return ["items": try reminders(c).filter { include || !$0.isCompleted }.map(reminderJSON)]
    }
    if op == "get" { return ["item": try type == .event ? eventJSON(event(r, c)) : reminderJSON(reminder(r, c))] }
    guard mutating else { throw Failure("Unknown operation") }
    guard r["apply"] as? Bool == true else { return ["state": "preview", "request": r] }
    let fields = r["fields"] as? [String: Any] ?? [:]
    let token = r["token"] as? String ?? ""
    let marker = "[steward:\(token)]"
    if op == "create", token.isEmpty { throw Failure("Create needs a stable token; reuse it when recovering an uncertain result") }
    if type == .event {
        let e: EKEvent
        if op == "create" {
            let start = try date(str(fields, "start")), end = try date(str(fields, "end"))
            guard end > start else { throw Failure("end must follow start") }
            let matches = store.events(matching: store.predicateForEvents(withStart: start, end: end, calendars: [c])).filter { ($0.notes ?? "").contains(marker) }
            if let existing = matches.first { return ["state": "existing", "item": eventJSON(existing)] }
            e = EKEvent(eventStore: store); e.calendar = c; e.startDate = start; e.endDate = end
            e.title = try str(fields, "title"); e.notes = marker; e.alarms = []
        } else { e = try event(r, c) }
        // Series edits and invitations need a dedicated UI-aware workflow.
        guard !e.hasRecurrenceRules, !e.isDetached, (e.attendees?.isEmpty ?? true), e.organizer == nil else { throw Failure("Recurring or invited event: use Calendar UI to choose occurrence/series and notification scope") }
        if op == "delete" {
            try store.remove(e, span: .thisEvent, commit: true)
            guard store.event(withIdentifier: try str(r, "id")) == nil else { throw Failure("Deletion read-back did not confirm removal") }
            return ["state": "deleted", "id": try str(r, "id")]
        }
        if let v = fields["title"] as? String { e.title = v }
        if let v = fields["start"] as? String { e.startDate = try date(v) }
        if let v = fields["end"] as? String { e.endDate = try date(v) }
        guard e.endDate > e.startDate else { throw Failure("end must follow start") }
        if let v = fields["notes"] as? String { e.notes = op == "create" ? v + "\n" + marker : v }
        if let v = fields["location"] as? String { e.location = v }
        if let v = fields["allDay"] as? Bool { e.isAllDay = v }
        if e.isAllDay {
            guard Calendar.current.startOfDay(for: e.startDate) == e.startDate,
                  Calendar.current.startOfDay(for: e.endDate) == e.endDate else {
                throw Failure("All-day boundaries must be midnight in the Mac's local timezone, with exclusive end")
            }
        }
        if let minutes = fields["alertMinutesBefore"] as? Int { e.alarms = [EKAlarm(relativeOffset: -Double(minutes * 60))] }
        else if fields["alertMinutesBefore"] is NSNull { e.alarms = [] }
        try store.save(e, span: .thisEvent, commit: true)
        guard let saved = store.event(withIdentifier: e.eventIdentifier) else { throw Failure("Saved but read-back failed; inspect before retry") }
        return ["state": "saved", "item": eventJSON(saved)]
    }
    let item: EKReminder
    if op == "create" {
        if let existing = try reminders(c).first(where: { ($0.notes ?? "").contains(marker) }) { return ["state": "existing", "item": reminderJSON(existing)] }
        item = EKReminder(eventStore: store); item.calendar = c; item.title = try str(fields, "title"); item.notes = marker
    } else { item = try reminder(r, c) }
    guard !item.hasRecurrenceRules else { throw Failure("Recurring reminder: use Reminders UI for series-aware edits") }
    if op == "delete" {
        try store.remove(item, commit: true)
        guard store.calendarItem(withIdentifier: try str(r, "id")) == nil else { throw Failure("Deletion read-back did not confirm removal") }
        return ["state": "deleted", "id": try str(r, "id")]
    }
    if let v = fields["title"] as? String { item.title = v }
    if let v = fields["notes"] as? String { item.notes = op == "create" ? v + "\n" + marker : v }
    if let v = fields["completed"] as? Bool { item.isCompleted = v }
    if let v = fields["priority"] as? Int { guard (0...9).contains(v) else { throw Failure("priority must be 0...9") }; item.priority = v }
    if let v = fields["due"] as? String {
        let d = try date(v)
        var dc = Calendar.current.dateComponents(in: TimeZone.current, from: d)
        dc.second = nil; dc.nanosecond = nil
        item.dueDateComponents = dc
    } else if fields["due"] is NSNull { item.dueDateComponents = nil }
    if let when = fields["remindAt"] as? String { item.alarms = [EKAlarm(absoluteDate: try date(when))] }
    else if fields["remindAt"] is NSNull { item.alarms = [] }
    try store.save(item, commit: true)
    guard let saved = store.calendarItem(withIdentifier: item.calendarItemIdentifier) as? EKReminder else { throw Failure("Saved but read-back failed; inspect before retry") }
    return ["state": "saved", "item": reminderJSON(saved)]
}
do {
    let data = FileHandle.standardInput.readDataToEndOfFile()
    guard let request = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw Failure("Expected JSON object") }
    emit(try run(request))
} catch { emit(["error": String(describing: error), "state": "failed-or-uncertain"]); exit(1) }

File: skills/steward-apple/scripts/apple.py
#!/usr/bin/env python3
"""Steward's local Apple Calendar/Reminders connector. JSON stdin -> JSON stdout.

Operations: status, connect, containers, list, get, create, update, delete,
create-container, rename-container, delete-container (empty only).
Container creation takes source and name; renaming takes calendar and name.
app: calendar|reminders|notes. calendar: discovered container ID.
Calendar list requires start/end ISO8601 timestamps (maximum 93 days).
Reminders list optionally includes includeCompleted:true.
Create requires token (stable UUID for retry recovery), fields.title, and for events
fields.start/end. Update/delete require id. Mutations preview unless --apply.
Fields: calendar title/start/end/notes/location/allDay; reminders
title/notes/due/completed/priority/remindAt. Calendar alertMinutesBefore sets
a notification alarm; null clears alarms. Reminder remindAt uses an explicit
timestamp; null clears alarms. Notes title/text/appendText; replacement text
requires expectedModified from get. due:null clears a due date. Dates need offsets.
Run --build to compile locally (macOS 14+, Apple developer tools required).
"""
import argparse
import hashlib
import json
import os
from pathlib import Path
import platform
import re
import fcntl
import subprocess
import sys
import tempfile
from datetime import datetime

HERE = Path(__file__).resolve().parent
CACHE = Path.home() / 'Library/Caches/Steward/apple'
WRITES = {'create', 'update', 'delete', 'create-container', 'rename-container', 'delete-container'}


def timestamp(value):
    if not isinstance(value, str) or not re.fullmatch(r'\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(Z|[+-]\d{2}:\d{2})', value):
        raise ValueError('Date must be a string with an explicit UTC offset')
    d = datetime.fromisoformat(value.replace('Z', '+00:00'))
    if d.tzinfo is None or d.microsecond:
        raise ValueError('Use ISO8601 seconds and explicit UTC offset')
    return d


def validate(r):
    if not isinstance(r, dict):
        raise ValueError('Expected JSON object')
    op = r.get('op')
    if op not in {'status', 'connect', 'containers', 'list', 'get', *WRITES}:
        raise ValueError('Unknown op')
    allowed = {'op', 'app', 'calendar', 'id', 'fields', 'token', 'start', 'end', 'includeCompleted', 'source', 'name', 'expectedModified'}
    if set(r) - allowed:
        raise ValueError('Unknown request keys: ' + ', '.join(sorted(set(r) - allowed)))
    if op == 'status':
        return r
    app = r.get('app')
    if app not in {'calendar', 'reminders', 'notes'}:
        raise ValueError('app must be calendar, reminders or notes')
    if op not in {'connect', 'containers', 'create-container'} and not isinstance(r.get('calendar'), str):
        raise ValueError('Choose a container by its discovered calendar ID')
    if op in {'get', 'update', 'delete'} and not r.get('id'):
        raise ValueError('id is required')
    if op == 'list' and app == 'calendar':
        start, end = timestamp(r.get('start')), timestamp(r.get('end'))
        if not 0 < (end - start).total_seconds() <= 93 * 86400:
            raise ValueError('Calendar query must cover 1 second to 93 days')
    if op in {'create-container', 'rename-container'}:
        if not isinstance(r.get('name'), str) or not r['name'].strip(): raise ValueError('Container name is required')
    if op == 'create-container' and not isinstance(r.get('source'), str): raise ValueError('Discovered source ID is required')
    f = r.get('fields', {})
    if not isinstance(f, dict):
        raise ValueError('fields must be an object')
    fields = {'title', 'notes', 'location', 'start', 'end', 'allDay', 'alertMinutesBefore'} if app == 'calendar' else ({'title', 'notes', 'due', 'completed', 'priority', 'remindAt'} if app == 'reminders' else {'title', 'text', 'appendText'})
    if set(f) - fields:
        raise ValueError('Unsupported fields: ' + ', '.join(sorted(set(f) - fields)))
    for k, v in f.items():
        if k in {'allDay', 'completed'}:
            if type(v) is not bool: raise ValueError(k + ' must be boolean')
        elif k == 'alertMinutesBefore':
            if v is not None and (type(v) is not int or not 0 <= v <= 40320): raise ValueError('alertMinutesBefore must be null or 0...40320')
        elif k == 'priority':
            if type(v) is not int or not 0 <= v <= 9: raise ValueError('priority must be 0...9')
        elif k in {'due', 'remindAt'} and v is None:
            pass
        elif not isinstance(v, str):
            raise ValueError(k + ' must be a string')
        if k in {'start', 'end', 'due', 'remindAt'} and v is not None:
            timestamp(v)
    if 'text' in f and 'appendText' in f: raise ValueError('Choose replacement text or appendText')
    if op == 'update' and app == 'notes' and not r.get('expectedModified'):
        raise ValueError('Read note first and supply expectedModified for any update')
    if op == 'create':
        if app == 'notes' and 'appendText' in f: raise ValueError('Use text for initial note content')
        import uuid
        if not isinstance(r.get('token'), str): raise ValueError('token must be a UUID string')
        uuid.UUID(r['token'])
        if not f.get('title', '').strip(): raise ValueError('title is required')
        if app == 'calendar' and not {'start', 'end'} <= set(f):
            raise ValueError('Calendar create requires start and end')
    if 'start' in f and 'end' in f and timestamp(f['end']) <= timestamp(f['start']):
        raise ValueError('end must follow start')
    if f.get('allDay'):
        for key in ('start', 'end'):
            if key in f and any((timestamp(f[key]).hour, timestamp(f[key]).minute, timestamp(f[key]).second)):
                raise ValueError('All-day boundaries must be local midnight, with exclusive end')
    if 'title' in f and not f['title'].strip(): raise ValueError('title cannot be blank')
    return r


def build():
    if platform.system() != 'Darwin': raise ValueError('This connector requires macOS')
    digest = hashlib.sha256((HERE / 'apple.swift').read_bytes() + (HERE / 'Info.plist').read_bytes()).hexdigest()
    binary = CACHE / 'steward-apple'
    stamp = CACHE / 'build.sha256'
    if binary.exists() and stamp.exists() and stamp.read_text() == digest:
        return binary
    CACHE.mkdir(parents=True, exist_ok=True)
    temp = CACHE / ('build-' + str(os.getpid()))
    compilers = [Path('/Library/Developer/CommandLineTools/usr/bin/swiftc'), Path('/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc')]
    errors = []
    for compiler in compilers:
        if not compiler.exists(): continue
        cmd = [str(compiler), '-target', platform.machine() + '-apple-macosx14.0', str(HERE / 'apple.swift'), '-o', str(temp), '-Xlinker', '-sectcreate', '-Xlinker', '__TEXT', '-Xlinker', '__info_plist', '-Xlinker', str(HERE / 'Info.plist')]
        sdk = Path('/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk') if str(compiler).startswith('/Applications/') else Path('/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk')
        if sdk.exists(): cmd += ['-sdk', str(sdk)]
        cmd += ['-module-cache-path', str(CACHE / 'modules')]
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=180)
        if result.returncode == 0:
            temp.replace(binary); stamp.write_text(digest); return binary
        errors.append(result.stderr[-3000:])
    raise ValueError('Could not compile. Install/repair Apple Command Line Tools.\n' + '\n'.join(errors))


def main():
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument('--apply', action='store_true', help='Execute the authorized mutation; default returns a local preview')
    p.add_argument('--build', action='store_true', help='Compile without requesting app access')
    a = p.parse_args()
    try:
        if a.build:
            print(json.dumps({'binary': str(build())})); return 0
        r = validate(json.load(sys.stdin))
        if r['op'] in WRITES and not a.apply:
            print(json.dumps({'state': 'preview', 'request': r})); return 0
        r['apply'] = a.apply
        CACHE.mkdir(parents=True, exist_ok=True)
        native_script = r.get('app') == 'notes'
        binary = None if native_script else build()
        with (CACHE / 'operation.lock').open('a') as lock:
            fcntl.flock(lock, fcntl.LOCK_EX)
            if native_script:
                with tempfile.NamedTemporaryFile(mode='w', suffix='.json') as request_file:
                    json.dump(r, request_file); request_file.flush()
                    result = subprocess.run(['/usr/bin/osascript', '-l', 'JavaScript', str(HERE / 'notes.js'), request_file.name], text=True, capture_output=True, timeout=100)
            else:
                result = subprocess.run([str(binary)], input=json.dumps(r), text=True, capture_output=True, timeout=100)
        if result.stdout: print(result.stdout.strip())
        if result.stderr: print(result.stderr.strip(), file=sys.stderr)
        if result.returncode: return result.returncode
        payload = json.loads(result.stdout)
        if r['op'] == 'create-container' and r.get('app') == 'notes' and payload.get('state') == 'created':
            c = payload['container']
            if not c['shared']:
                policy_path = Path.home() / 'Library/Application Support/Steward/apple-policy.json'
                policy = json.loads(policy_path.read_text())
                policy.setdefault('notes', {})[c['id']] = {'name': c['name'], 'write': True, 'provenance': 'created-private-by-Steward'}
                temp = policy_path.with_suffix('.tmp')
                temp.write_text(json.dumps(policy, indent=2)); temp.chmod(0o600); temp.replace(policy_path)
        return 1 if 'error' in payload else 0
    except (ValueError, OSError, subprocess.SubprocessError) as e:
        print(json.dumps({'error': str(e), 'state': 'failed-or-uncertain', 'recovery': 'Read back the destination before retrying a mutation; do not use a new create token.'}))
        return 1


if __name__ == '__main__':
    sys.exit(main())
